### ISSUE TITLE
Enable operation on groups of upstream packages

### ISSUE THREAD

**Issue by johnbelamaric** [nephio-project/nephio#336]
Currently, PackageVariant and PackageVariantSet resources operate on a single upstream package. We should consider extending that to multiple upstream packages that are require the same mutations and targeting.

/sig automation
/area package-management

**Comment by liamfallon** [comment]
Triaged

**Comment by liamfallon** [comment]
triaged

**Comment by thc1006** [comment]
Dear @liamfallon ,

I would like to work on this issue.

I have reviewed the PackageVariant and PackageVariantSet codebase and understand the scope of changes needed to support multiple upstream packages.

Since this issue has the design label, I plan to submit a design document for review before starting implementation.

May I assign myself to this issue?

**Comment by thc1006** [comment]
Following up with a more detailed analysis.

## Background Research

I found that the original design document ([08-package-variant.md](https://github.com/GoogleContainerTools/kpt/blob/main/docs/design-docs/08-package-variant.md)) already discussed this in the **"Future Work"** section:

> "Installing a collection of packages across a set of clusters, or performing the same mutations to each package in a collection, is only supported by creating multiple PackageVariant / PackageVariantSet resources."

Three options were proposed:
1. An `upstreams` field listing multiple packages
2. A `PackageRevisionSet` resource with a reusable list of upstreams
3. Listing multiple PackageRevisionSets in Upstream

## Current Architecture Constraints

| Component | File | Line | Constraint |
|-----------|------|------|------------|
| PackageVariant | `packagevariant_types.go` | 59 | `Upstream *Upstream` (single pointer) |
| PackageVariantSet | `packagevariantset_types.go` | 47 | `Upstream *pkgvarapi.Upstream` (single pointer) |
| PV Controller | `packagevariant_controller.go` | 266 | `getUpstreamPR()` returns single PR |
| PVS Controller | `packagevariantset_controller.go` | 162 | `getUpstreamPR()` returns single PR |

## Use Cases

This would benefit real-world scenarios like Nephio's Free5GC deployment:
- `pkg-example-amf-bp`, `pkg-example-smf-bp`, `pkg-example-upf-bp` are separate packages
- They often require identical mutations and targeting
- Currently requires creating 3 separate PackageVariantSet resources

## Proposed Design Options

Building on the original design doc's suggestions and existing patterns in the codebase:

### Option A: Upstreams List (from original design doc)
```go
type PackageVariantSetSpec struct {
    Upstreams []Upstream `json:"upstreams,omitempty"`  // NEW
    Upstream  *Upstream  `json:"upstream,omitempty"`   // Keep for compatibility
}
```
**Pros**: Directly implements the original design intent, backwards compatible
**Cons**: Requires looping logic in reconciler

### Option B: Upstream Selector (follows existing patterns)
```go
type PackageVariantSetSpec struct {
    UpstreamSelector *UpstreamSelector `json:"upstreamSelector,omitempty"`
}

type UpstreamSelector struct {
    Repo         string                `json:"repo,omitempty"`
    LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
}
```
**Pros**: Consistent with existing `RepositorySelector` and `ObjectSelector` patterns
**Cons**: Requires packages to have appropriate labels

### Option C: Package Name Pattern
```go
type Upstream struct {
    Repo           string `json:"repo,omitempty"`
    PackagePattern string `json:"packagePattern,omitempty"`  // e.g., "pkg-example-*-bp"
}
```
**Pros**: Simple glob matching for related packages
**Cons**: No existing pattern in codebase, less flexible

## Questions for Maintainers

1. **Design preference**: Which option aligns best with the project's direction? Or should I include all options in the design doc for discussion?
2. **Scope**: Should this apply to PackageVariant, PackageVariantSet, or both?
3. **API versioning**: Backwards-compatible extension to v1alpha1/v1alpha2, or new version?

## Next Steps

I'm starting work on a formal design document that will include:
- Detailed API changes for the chosen approach
- Controller logic modifications
- Migration/compatibility considerations
- Test plan

Will submit a PR to `docs/design/` within the next few days.

/assign
